src/util/submit-addon.js

```diff
@@ -0,0 +1,333 @@
+              clearTimeout(abortTimeout);
+              resolve(detailResponseData.uuid);
+            } else {
+              log.info('Validation failed.');
+              clearTimeout(abortTimeout);
+
+              reject(new Error(detailResponseData.url));
+            }
+          } else {
+            // Validation is still in progress, so wait for a while and try again.
+            validationCheckTimeout = setTimeout(
+              pollValidationStatus,
+              this.validationCheckInterval
+            );
+          }
+        } catch (err) {
+          clearTimeout(abortTimeout);
+          reject(err);
+        }
+      };
+
+      pollValidationStatus();
+    });
+  }
+
+  async doNewAddonSubmit(uuid: string, metaDataJson: Object): Promise<any> {
+    const url = `${this.apiUrl}addon/`;
+    const jsonData = { version: { upload: uuid }, ...metaDataJson };
+    return this.fetchJson(url, 'POST', JSON.stringify(jsonData));
+  }
+
+  doNewAddonOrVersionSubmit(
+    addonId: string,
+    uuid: string,
+    metaDataJson: Object,
+  ): Promise<typeof Response> {
+    const url = `${this.apiUrl}addon/${addonId}/`;
+    const jsonData = { version: { upload: uuid }, ...metaDataJson };
+    return this.fetch(url, 'PUT', JSON.stringify(jsonData));
+  }
+
+  waitForApproval(addonId: string, versionId: number): Promise<string> {
+    let approvalCheckTimeout;
+    const detailUrl = `${this.apiUrl}addon/${addonId}/versions/${versionId}/`;
+
+    return new Promise((resolve, reject) => {
+      const abortTimeout = setTimeout(() => {
+        clearTimeout(approvalCheckTimeout);
+
+        reject(new Error('Approval Timeout.'));
+      }, this.approvalCheckTimeout);
+
+      const pollApprovalStatus = async () => {
+        try {
+          const detailResponseData = await this.fetchJson(
+            detailUrl, 'GET', undefined, 'Getting addon details failed.');
+
+          const file = detailResponseData.file;
+          if (file && file.status === 'public') {
+            clearTimeout(abortTimeout);
+            resolve(file.url);
+          } else {
+            // The add-on hasn't been approved yet, so wait for a while and try again.
+            approvalCheckTimeout = setTimeout(
+              pollApprovalStatus,
+              this.approvalCheckInterval
+            );
+          }
+        } catch (err) {
+          clearTimeout(abortTimeout);
+          reject(err);
+        }
+      };
+
+      pollApprovalStatus();
+    });
+  }
+
+  async fetchJson(
+    url: string,
+    method: string = 'GET',
+    body?: typeof FormData | string,
+    errorMsg: string = 'Bad Request'
+  ): Promise<any> {
+    const response = await this.fetch(url, method, body);
+    if (response.status < 100 || response.status >= 500) {
+      throw new Error(
+        `${errorMsg}: ${response.statusText || response.status}.`);
+    }
+    const data = await response.json();
+
+    if (!response.ok) {
+      log.info(data);
```

Comment:
I don't know what context the logger usually has but maybe we should add a message in addition to the object here.